AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1491,12 +1510,93 @@ - (void)measureNodeWithBoundsIfNecessary:(CGRect)bounds
   
   // If no measure pass happened or the bounds changed between layout passes we manually trigger a measurement pass
   // for the node using a size range equal to whatever bounds were provided to the node
-  if (supportsRangeManagedInterfaceState == NO && (hasDirtyLayout || CGSizeEqualToSize(calculatedLayoutSize, bounds.size) == NO)) {
-    if (CGRectEqualToRect(bounds, CGRectZero)) {
-      LOG(@"Warning: No size given for node before node was trying to layout itself: %@. Please provide a frame for the node.", self);
-    } else {
-      [self layoutThatFits:ASSizeRangeMake(bounds.size)];
+  if (CGRectEqualToRect(bounds, CGRectZero)) {
+    LOG(@"Warning: No size given for node before node was trying to layout itself: %@. Please provide a frame for the node.", self);
+    return;
+  }
+
+  
+  // Check if we can reuse the calculated display node layout
+  if (_pendingDisplayNodeLayout == nullptr &&
+      _calculatedDisplayNodeLayout->isDirty() == NO &&
+      CGSizeEqualToSize(_calculatedDisplayNodeLayout->layout.size, bounds.size))
+  {
+    // Reuse calculatedDisplayNodeLayout for layout pass
+    return;
+  }
+    
+  // Calcualted layout is not reusable we need to transform to a new one
+  [self cancelLayoutTransition];
+    
+  BOOL didCreateNewContext = NO;
+  BOOL didOverrideExistingContext = NO;
+  BOOL shouldVisualizeLayout = ASHierarchyStateIncludesVisualizeLayout(_hierarchyState);
+  ASLayoutElementContext context;
+  if (ASLayoutElementContextIsNull(ASLayoutElementGetCurrentContext())) {
+    context = ASLayoutElementContextMake(ASLayoutElementContextDefaultTransitionID, shouldVisualizeLayout);
+    ASLayoutElementSetCurrentContext(context);
+    didCreateNewContext = YES;
+  } else {
+    context = ASLayoutElementGetCurrentContext();
+    if (context.needsVisualizeNode != shouldVisualizeLayout) {
+      context.needsVisualizeNode = shouldVisualizeLayout;
+      ASLayoutElementSetCurrentContext(context);
+      didOverrideExistingContext = YES;
+    }
+  }
+  
+  // Figure out previos and pending layout for layout transition
+  auto previousLayout = _calculatedDisplayNodeLayout;
+  auto pendingLayout = [=]() -> std::shared_ptr<ASDisplayNodeLayout> {
+    // Check if  the pending display node layout can be used to transition to
+    if (_pendingDisplayNodeLayout != nullptr &&
+        _pendingDisplayNodeLayout->isDirty() == NO &&
+        CGSizeEqualToSize(_pendingDisplayNodeLayout->layout.size, bounds.size)) {
+      return _pendingDisplayNodeLayout;
+    }
+  
+    CGSize parentSize = bounds.size;
+    ASSizeRange constrainedSize = ASSizeRangeMake(parentSize);
+    
+    // Checkout if constrained size of layouts can be reused
+    if (_pendingDisplayNodeLayout != nullptr) {
+      constrainedSize = _pendingDisplayNodeLayout->constrainedSize;
+    } else if (CGSizeEqualToSize(_calculatedDisplayNodeLayout->layout.size, CGSizeZero) == NO) {
+      constrainedSize = _calculatedDisplayNodeLayout->constrainedSize;
```

Comment:
This is the good logic that we should share any method so that the other places that need to act says the current constrain size will be able to use it. I think the only change needed here is that it should fall back to the bounds in the worst case?